### PR TITLE
Add Lifetime Encoding for Function Calls

### PR DESCRIPTION
**Pull request by pascal-huber**
On caller side:
 - [x] Assert Lifetime preconditions
 - [x] Exhale Lifetimes access (function lifetime,  argument lifetimes, static lifetime)
 - [x] Inhale Lifetimes again after function call
 
On called side:
  - [x] Inhale Lifetime access
  - [x] Inhale Lifetime preconditions

### CODE REVIEW

File: vir/defs/high/ast/predicate.rs [hunk 1]
```diff
@@ -1,19 +1,27 @@
...
 pub enum Predicate {
+    LifetimeToken(LifetimeToken),
     MemoryBlockStack(MemoryBlockStack),
     MemoryBlockStackDrop(MemoryBlockStackDrop),
     MemoryBlockHeap(MemoryBlockHeap),
     MemoryBlockHeapDrop(MemoryBlockHeapDrop),
     OwnedNonAliased(OwnedNonAliased),
 }
 
+#[display(fmt = "LifetimeToken")]
```
Comment (vakaras) [inline vir/defs/high/ast/predicate.rs]:
```suggestion
#[display(fmt = "acc(LifetimeToken({}), {})", lifetime, rd_perm)]
```

Comment (pascal-huber) [reply vir/defs/high/ast/predicate.rs]: done

File: vir/defs/middle/ast/statement.rs [hunk 2]
```diff
@@ -267,6 +268,14 @@ pub struct LifetimeReturn {
     pub position: Position,
 }
 
+#[display(fmt = "lifetime_included({}, {})", lhs, "display::cjoin(rhs)")]
```
Comment (vakaras) [inline vir/defs/middle/ast/statement.rs]:
```suggestion
// TODO: Make LifetimeIncluded an Expression and use Assert/Assume Statement
// We should use BuiltinFuncApp for this: https://github.com/viperproject/prusti-dev/blob/master/vir/defs/high/ast/expression.rs#L245-L252
```

File: vir/defs/high/ast/statement.rs [hunk 3]
```diff
@@ -271,6 +272,15 @@ pub struct LifetimeReturn {
     pub position: Position,
 }
 
+// TODO: Make LifetimeIncluded an Expression and use Assert/Assume Statement
```
Comment (vakaras) [inline vir/defs/high/ast/statement.rs]:
```suggestion
// TODO: Make LifetimeIncluded an Expression and use Assert/Assume Statement
// We should use BuiltinFuncApp for this: https://github.com/viperproject/prusti-dev/blob/master/vir/defs/high/ast/expression.rs#L245-L252
```

File: prusti-viper/src/encoder/mir/procedures/encoder/scc.rs [hunk 4]
```diff
@@ -0,0 +1,126 @@
+// source: https://rosettacode.org/wiki/Tarjan#Rust
```
Comment (vakaras) [inline prusti-viper/src/encoder/mir/procedures/encoder/scc.rs:1]: Please add a comment mentioning the license.

File: prusti-viper/src/encoder/high/procedures/interface.rs [hunk 5]
```diff
@@ -22,6 +22,7 @@ impl<'v, 'tcx: 'v> HighProcedureEncoderInterface<'tcx> for super::super::super::
         debug!("procedure_high:\n{}", procedure_high);
         let procedure =
             super::inference::infer_shape_operations(self, proc_def_id, procedure_high)?;
+        // println!("end of encode_procedure_core_proof");
```
Comment (vakaras) [inline prusti-viper/src/encoder/high/procedures/interface.rs]:
```suggestion
```

File: prusti-viper/src/encoder/middle/core_proof/lifetimes/interface.rs [hunk 6]
```diff
@@ -262,6 +266,26 @@ impl<'p, 'v: 'p, 'tcx: 'v> LifetimesInterface for Lowerer<'p, 'v, 'tcx> {
...
+        lifetimes: &[vir_low::VariableDecl],
+    ) -> SpannedEncodingResult<vir_low::Expression> {
+        self.encode_lifetime_token_predicate()?;
+        self.encode_lifetime_intersect(lifetimes.len())?;
+        use vir_low::macros::*;
+        let mut arguments: Vec<vir_low::Expression> = Vec::new();
+        for lifetime in lifetimes {
+            arguments.push(lifetime.clone().into());
+        }
+        let function_call = vir_low::Expression::domain_function_call(
```
Comment (vakaras) [inline prusti-viper/src/encoder/middle/core_proof/lifetimes/interface.rs]: This n-ary function call looks strange: what axioms do you use to define it? What is the advantage compared to using the binary intersection function?

File: prusti-viper/src/encoder/mir/procedures/encoder/lifetimes.rs [hunk 7]
```diff
@@ -391,4 +448,219 @@ impl<'p, 'v: 'p, 'tcx: 'v> LifetimesEncoder for ProcedureEncoder<'p, 'v, 'tcx> {
...
+        };
+
+        // Precondition: Inhale LifetimeTokens
+        let mut preconditions = vec![vir_high::Statement::comment(
+            "Lifetime preconditions.".to_string(),
+        )];
+        let lifetimes_to_inhale: BTreeSet<vir_high::ty::LifetimeConst> =
+            self.lifetimes_to_inhale()?;
+        for lifetime in &lifetimes_to_inhale {
+            let inhale_statement = self.encode_inhale_lifetime_token(lifetime.clone(), 1)?;
```
Comment (vakaras) [inline prusti-viper/src/encoder/mir/procedures/encoder/lifetimes.rs]:
```suggestion
            let inhale_statement = self.encode_inhale_lifetime_token(lifetime.clone(), rd)?;
```
This should be not `1`, but some unknown positive permission amount. You can use an unconstrained variable of type [`MPerm`](https://github.com/viperproject/prusti-dev/blob/15f651bc3cfffd78583b3862f2be9e0ad9b4feee/vir/defs/high/ast/ty.rs#L14-L15) for this.

File: prusti-viper/src/encoder/mir/procedures/encoder/mod.rs [hunk 8]
```diff
@@ -22,6 +22,7 @@ use crate::encoder::{
     mir_encoder::PRECONDITION_LABEL,
     Encoder,
 };
+
```
Comment (vakaras) [inline prusti-viper/src/encoder/mir/procedures/encoder/mod.rs]:
```suggestion
```

File: prusti-viper/src/encoder/mir/procedures/encoder/mod.rs [hunk 9]
```diff
@@ -84,6 +86,7 @@ pub(super) fn encode_procedure<'v, 'tcx: 'v>(
     let specification_blocks = SpecificationBlocks::build(tcx, mir, &procedure);
     let initialization = compute_definitely_initialized(def_id, mir, encoder.env().tcx());
     let allocation = compute_definitely_allocated(def_id, mir);
+    let function_call_ctr: &mut usize = &mut 0;
```
Comment (vakaras) [inline prusti-viper/src/encoder/mir/procedures/encoder/mod.rs]: Why is this a mutable reference?

**Review by vakaras — CHANGES_REQUESTED** [review]
Overall looks good, but before merging I would like two changes to be made:

1. Lifetime tokens inside a function should be inhaled not with a full lifetime, but with some unknown positive permission amount `q`. This will require to use that permission amount when encoding all `rd_perm`s: instead of `acc(..., 1/rd_perm)` it should be `acc(..., q/rd_perm)`.
2. Instead of `LifetimeIncluded` statement it should be an expression `included(lft1, intersect(lft2, lft3))` or something like that.

File: prusti-viper/src/encoder/mir/procedures/encoder/mod.rs (hunk 9 above)
Comment (pascal-huber) [reply prusti-viper/src/encoder/mir/procedures/encoder/mod.rs]: No reason - changed it to `let function_call_ctr: usize = 0;`.

File: prusti-viper/src/encoder/mir/procedures/encoder/mod.rs (hunk 8 above)
Comment (pascal-huber) [reply prusti-viper/src/encoder/mir/procedures/encoder/mod.rs]: done

File: prusti-viper/src/encoder/high/procedures/interface.rs (hunk 5 above)
Comment (pascal-huber) [reply prusti-viper/src/encoder/high/procedures/interface.rs]: ups, done

File: prusti-viper/src/encoder/middle/core_proof/lifetimes/interface.rs (hunk 6 above)
Comment (pascal-huber) [reply prusti-viper/src/encoder/middle/core_proof/lifetimes/interface.rs]: The idea was that if I construct a binary operation here, I will have to destruct it again later to create the function call. It should translate to the domain function `intersect`.

File: prusti-viper/src/encoder/mir/procedures/encoder/scc.rs (hunk 4 above)
Comment (pascal-huber) [reply prusti-viper/src/encoder/mir/procedures/encoder/scc.rs:1]: done

File: vir/defs/high/ast/statement.rs (hunk 3 above)
Comment (pascal-huber) [reply vir/defs/high/ast/statement.rs]: added comment

File: vir/defs/middle/ast/statement.rs (hunk 2 above)
Comment (pascal-huber) [reply vir/defs/middle/ast/statement.rs]: I assume removing `fmt` in the suggestion is a mistake? I added the comments.

**Comment by pascal-huber** [comment]
> Overall looks good, but before merging I would like two changes to be made:
> 
>     1. Lifetime tokens inside a function should be inhaled not with a full lifetime, but with some unknown positive permission amount `q`. This will require to use that permission amount when encoding all `rd_perm`s: instead of `acc(..., 1/rd_perm)` it should be `acc(..., q/rd_perm)`.

Ok, can do this. Is there an advantage of using an unconstrained permission amount?

> 
>     2. Instead of `LifetimeIncluded` statement it should be an expression `included(lft1, intersect(lft2, lft3))` or something like that.

Currently this is done in `into_low/cfg.rs`. It produces for example:

```
  inhale included$(lft_0$snapshot$0, intersect$3(lft_1$snapshot$0, lft_2$snapshot$0, lft_3$snapshot$0))
```

File: prusti-viper/src/encoder/middle/core_proof/lifetimes/interface.rs (hunk 6 above)
Comment (pascal-huber) [reply prusti-viper/src/encoder/middle/core_proof/lifetimes/interface.rs]: solved by using `BuiltinFuncApp`

File: prusti-viper/src/encoder/mir/procedures/encoder/lifetimes.rs (hunk 7 above)
Comment (pascal-huber) [reply prusti-viper/src/encoder/mir/procedures/encoder/lifetimes.rs]: done

File: vir/defs/high/ast/expression.rs [hunk 10]
```diff
@@ -259,3 +261,12 @@ pub struct Downcast {
     pub field: FieldDecl,
     pub position: Position,
 }
+
+impl Expression {
```
Comment (vakaras) [inline vir/defs/high/ast/expression.rs]: This should be in [this file](https://github.com/viperproject/prusti-dev/blob/master/vir/defs/high/operations_internal/expression.rs).

File: vir/src/low/operations/ty.rs [hunk 11]
```diff
@@ -159,6 +159,8 @@ impl Typed for BinaryOp {
             | BinaryOpKind::Mod => {
                 let ty1 = self.left.get_type();
                 let ty2 = self.right.get_type();
+                println!("-- get_type --");
```
Comment (vakaras) [inline vir/src/low/operations/ty.rs]: Please remove debug output.
```suggestion
```

**Review by vakaras — APPROVED** [review]
Overall looks good. Once the two minor things are addressed, I will merge.

File: vir/src/low/operations/ty.rs (hunk 11 above)
Comment (pascal-huber) [reply vir/src/low/operations/ty.rs]: missed that one, done

File: vir/defs/high/ast/expression.rs (hunk 10 above)
Comment (pascal-huber) [reply vir/defs/high/ast/expression.rs]: moved

**Review by vakaras — APPROVED** [review]
bors r+

**Comment by vakaras** [comment]
bors retry

**Comment by vakaras** [comment]
bors cancel

**Comment by vakaras** [comment]
bors cancel

**Pull request merged**